### PR TITLE
[storage] Change generated code Uri.ToString to Uri.AbsoluteUri

### LINKED ISSUE #11368: [BUG] Storage Blobs - can't copy from URL with non-ASCII characters
**Describe the bug**
StageBlockFromUri, StartCopyFromUri, etc. fail when the source URI contains non-ASCII characters

**Expected behavior**
They should work with source URIs with non-ASCII characters.

According to https://docs.microsoft.com/en-us/rest/api/storageservices/put-block-from-url the `x-ms-copy-source` header "should be URL-encoded as it would appear in a request URI".

The previous SDK handled non-ASCII characters by using the Uri.AbsoluteUri property, which encodes non-ASCII characters. (See https://github.com/Azure/azure-storage-net/blob/master/Lib/ClassLibraryCommon/Blob/Protocol/BlobHttpRequestMessageFactory.cs#L260)

**Actual behavior (include Exception or Stack Trace)**
BlobRestClient.StageBlockFromUriAsync_CreateMessage calls Uri.ToString() (rather that AbsoluteUri), which is not encoded. Even if the original string passed to the Uri constructor escapes the non-ASCII characters, .ToString() canonicalizes and un-escapes them. The non-ASCII characters get put in the x-ms-copy-source header un-escaped, so HttpClient throws an HttpRequestException that "Request headers must contain only ASCII characters."

```
System.AggregateException
  HResult=0x80131500
  Message=Retry failed after 6 tries. (Request headers must contain only ASCII characters.) (Request headers must contain only ASCII characters.) (Request headers must contain only ASCII characters.) (Request headers must contain only ASCII characters.) (Request headers must contain only ASCII characters.) (Request headers must contain only ASCII characters.)
  Source=Azure.Core
  StackTrace:
(Omitted for brevity; the AggregateException trace doesn't point to the actual problem.)

Inner Exception 1:
RequestFailedException: Request headers must contain only ASCII characters.

Inner Exception 2:
HttpRequestException: Request headers must contain only ASCII characters.

```

**To Reproduce**
```
class Program
{
	static void Main(string[] args)
	{
		var sourceUri = new System.Uri("https://via.placeholder.com/200.png?foo=bar&text=hello world…‽");

		// Note: even encoding the URL as per below fails, because Uri.ToString() canonicalizes it and decodes it
		// var sourceUri = new System.Uri("https://via.placeholder.com/200.png?foo=bar&text=hello%20world%E2%80%A6%E2%80%BD");


		var destinationClient = new Azure.Storage.Blobs.Specialized.BlockBlobClient("UseDevelopmentStorage=true", "test", "output.png");
		destinationClient.StageBlockFromUri(sourceUri, System.Convert.ToBase64String(new byte[] { 1, 2, 3 }));

		// Also fails in a similar way:
		//destinationClient.StartCopyFromUri(sourceUri);


		// Old Microsoft.Azure.Storage.Blob SDK doesn't fail:

		//var oldSdkClient = Microsoft.Azure.Storage.Blob.BlobAccountExtensions.CreateCloudBlobClient(Microsoft.Azure.Storage.CloudStorageAccount.DevelopmentStorageAccount);
		//var oldSdkCloudBlockBlob = oldSdkClient.GetContainerReference("test").GetBlockBlobReference("old.png");
		//oldSdkCloudBlockBlob.StartCopy(sourceUri);
	}
}
```

**Environment:**
 - Name and version of the Library package used: Azure.Storage.Blobs 12.4.1. (Regression compared to Microsoft.Azure.Storage.Blob 11.1.3.)
 - Hosting platform or OS and .NET runtime version (`dotnet --info` output for .NET Core projects): 
.NET Core SDK (reflecting any global.json):
 Version:   3.1.201
 Commit:    b1768b4ae7

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19041
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.201\

Host (useful for support):
  Version: 3.1.3
  Commit:  4a9f85e9f8

 - IDE and version : Visual Studio 16.5.4

### PR DESCRIPTION
**Pull request by amnguye**
Resolves #11368

Uri.ToString() unencodes the Uri before sending the request to storage. When using a source Uri that has non ascii characters, we will encode the Uri by putting it in the Uri object but unencode it when setting the header.

Switching to using Uri.AbsoluteUri fixes this problem. This fix applies to AppendBlobClient.AppendBlockFromUri. BlockBlobClient.StageBlockFromUri, PageBlobClient.UploadPagesFromUri, FileClient.StartCopy, FileClient.UploadRangeFromUri

This PR also has updates to DataLakeErrorDetails because of the regenerated code.

- Updated generator to convert Uri.ToString to Uri.AbsoluteUri
- Regenerated adding in Error to DataLakeErrorDetails
- Added tests and recordings for non Ascii source blob/file names

### CODE REVIEW

**Review by seanmcc-msft — COMMENTED** [review]
Looks good.

Why did we make the Data Lake error change?

**Comment by amnguye** [comment]
> Looks good.
> 
> Why did we make the Data Lake error change?

Hm. So I assumed someone from our team made the change to the storage swagger and we just haven't regenerated recently, but looks like the change was made by this PR

https://github.com/Azure/azure-rest-api-specs/pull/8739

Should I make a separate PR to make updates to the regenerated code?

**Comment by kasobol-msft** [comment]
> > Looks good.
> > Why did we make the Data Lake error change?
> 
> Hm. So I assumed someone from our team made the change to the storage swagger and we just haven't regenerated recently, but looks like the change was made by this PR
> 
> [Azure/azure-rest-api-specs#8739](https://github.com/Azure/azure-rest-api-specs/pull/8739)
> 
> Should I make a separate PR to make updates to the regenerated code?

We shouldn't be touching generated code (otherwise process would become non-deterministic at some point). If we were to change this we'd edit readme with generator transforms. However, this is just internal type rename, so we're good.

On the side note. https://github.com/Azure/azure-rest-api-specs/pull/8739/files it seems that we don't have this in 2019-12-12 rest spec. Would be good to get it there as well.

**Review by kasobol-msft — APPROVED** [review]

**Review by seanmcc-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs [hunk 1]
```diff
@@ -9225,7 +9225,7 @@ public static partial class PageBlob
 
                 // Add request headers
                 _request.Headers.SetValue("x-ms-version", version);
-                if (prevSnapshotUrl != null) { _request.Headers.SetValue("x-ms-previous-snapshot-url", prevSnapshotUrl.ToString()); }
+                if (prevSnapshotUrl != null) { _request.Headers.SetValue("x-ms-previous-snapshot-url", prevSnapshotUrl.AbsoluteUri); }
                 if (range != null) { _request.Headers.SetValue("x-ms-range", range); }
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs:9229]: Nit - Might be worth adding a test for snapshots to validate we're treating this URI correctly as well.

**Review by tg-msft — APPROVED** [review]

**Comment by tg-msft** [comment]
Yes - we should make the same swagger change to 2019-12-12 as @kasobol-msft noted.  The new Track 2 generators being developed are a lot pickier about correct swagger syntax instead of silently swallowing issues.  It's a good thing overall, but will require small changes like this.

**Pull request merged**